Source/Core/VideoCommon/VertexLoaderX64.cpp

```diff
@@ -0,0 +1,453 @@
+#ifdef _MSC_VER
+#include <intrin.h>
+#else
+#include <x86intrin.h>
+#endif
+
+#include "Common/CPUDetect.h"
+#include "Common/x64ABI.h"
+#include "VideoCommon/VertexLoaderX64.h"
+
+using namespace Gen;
+
+static const X64Reg src_reg = ABI_PARAM1;
+static const X64Reg dst_reg = ABI_PARAM2;
+static const X64Reg scratch1 = RAX;
+static const X64Reg scratch2 = ABI_PARAM3;
+static const X64Reg scratch3 = ABI_PARAM4;
+static const X64Reg count_reg = R10;
+static const X64Reg skipped_reg = R11;
+
+VertexLoaderX64::VertexLoaderX64(const TVtxDesc& vtx_desc, const VAT& vtx_att): VertexLoaderBase(vtx_desc, vtx_att)
+{
+	if (!IsInitialized())
+		return;
+
+	AllocCodeSpace(4096);
+	ClearCodeSpace();
+	GenerateVertexLoader();
+	WriteProtect();
+}
+
+OpArg VertexLoaderX64::GetVertexAddr(int array, u64 attribute)
+{
+	OpArg data = MDisp(src_reg, m_src_ofs);
+	if (attribute & MASK_INDEXED)
+	{
+		if (attribute == INDEX8)
+		{
+			MOVZX(64, 8, scratch1, data);
+			m_src_ofs += 1;
+		}
+		else
+		{
+			MOV(16, R(scratch1), data);
+			m_src_ofs += 2;
+			BSWAP(16, scratch1);
+			MOVZX(64, 16, scratch1, R(scratch1));
+		}
+		if (array == ARRAY_POSITION)
+		{
+			CMP(attribute == INDEX8 ? 8 : 16, R(scratch1), Imm8(-1));
+			m_skip_vertex = J_CC(CC_E, true);
+		}
+		MOVZX(64, 32, scratch2, M(&g_main_cp_state.array_strides[array]));
+		IMUL(64, scratch1, R(scratch2));
```

Comment:
Do we really need a 64-bit multiply here? I kinda doubt the array strides can be that large given the size of the GC memory space...

If the answer is yes, we really do need a 64-bit multiply, could you change array_strides to a u64 or ptrdiff_t to turn this into an IMUL with memory, saving an op? Otherwise you can just merge the two without doing that, if it's a 32-bit multiply.